gax/src/main/java/com/google/api/gax/rpc/mtls/MtlsProvider.java

```diff
@@ -0,0 +1,191 @@
+ * (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
+ * OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+ */
+
+package com.google.api.gax.rpc.mtls;
+
+import com.google.api.client.json.JsonParser;
+import com.google.api.client.json.gson.GsonFactory;
+import com.google.api.client.util.SecurityUtils;
+import com.google.api.core.BetaApi;
+import com.google.common.annotations.VisibleForTesting;
+import com.google.common.collect.ImmutableList;
+import java.io.FileInputStream;
+import java.io.FileNotFoundException;
+import java.io.IOException;
+import java.io.InputStream;
+import java.security.GeneralSecurityException;
+import java.security.KeyStore;
+import java.util.List;
+
+/**
+ * Provider class for mutual TLS. It is used to configure the mutual TLS in the transport with the
+ * default client certificate on device.
+ */
+@BetaApi
+public class MtlsProvider {
+  interface EnvironmentProvider {
+    String getenv(String name);
+  }
+
+  static class SystemEnvironmentProvider implements EnvironmentProvider {
+    @Override
+    public String getenv(String name) {
+      return System.getenv(name);
+    }
+  }
+
+  interface ProcessProvider {
+    public Process createProcess(InputStream metadata) throws IOException;
+  }
+
+  static class DefaultProcessProvider implements ProcessProvider {
+    @Override
+    public Process createProcess(InputStream metadata) throws IOException {
+      if (metadata == null) {
+        return null;
+      }
+      List<String> command = extractCertificateProviderCommand(metadata);
+      return new ProcessBuilder(command).start();
+    }
+  }
+
+  private static final String DEFAULT_CONTEXT_AWARE_METADATA_PATH =
+      System.getProperty("user.home") + "/.secureConnect/context_aware_metadata.json";
+
+  private String metadataPath;
+  private EnvironmentProvider envProvider;
+  private ProcessProvider processProvider;
+
+  public enum UseMtlsEndpoint {
+    NEVER,
+    AUTO,
+    ALWAYS;
+  }
+
+  @VisibleForTesting
+  MtlsProvider(
+      EnvironmentProvider envProvider, ProcessProvider processProvider, String metadataPath) {
+    this.envProvider = envProvider;
+    this.processProvider = processProvider;
+    this.metadataPath = metadataPath;
+  }
+
+  public MtlsProvider() {
+    this(
+        new SystemEnvironmentProvider(),
+        new DefaultProcessProvider(),
+        DEFAULT_CONTEXT_AWARE_METADATA_PATH);
+  }
+
+  /**
+   * Returns if mutual TLS client certificate should be used. If the value is true, the key store
+   * from {@link #getKeyStore()} will be used to configure mutual TLS transport.
+   */
+  public boolean useMtlsClientCertificate() {
+    String useClientCertificate = envProvider.getenv("GOOGLE_API_USE_CLIENT_CERTIFICATE");
+    return "true".equals(useClientCertificate);
+  }
+
+  /**
+   * Returns AUTO, NEVER or ALWAYS. NEVER means always use regular endpoint; ALWAYS means always use
```

Comment:
Consider moving this enum-value explanation into the enum itself, then this method comment can instead specify the context of UseMtlsEndpoint it returns (e.g. the current MTLS endpoint policy).

Reply:
done